File: pset6/crack/crack.py
#
#   Cracks passwords hashed with DES algorithm
#   works with up to five character password length
#

import sys
import crypt


def main():
    # check for correct argument input
    if len(sys.argv) != 2 or len(sys.argv[1]) < 1:
        print("Usage: python crack.py hash")
        exit(1)
    else:
        alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
        given_hash = sys.argv[1]
        salt = given_hash[:2]
        guess = [0]

        for fifth_letter in range(len(alphabet)):
            for fourth_letter in range(len(alphabet)):
                for third_letter in range(len(alphabet)):
                    for second_letter in range(len(alphabet)):
                        for first_letter in range(len(alphabet)):
                            guess[0] = alphabet[first_letter]
                            crypted = crypt.crypt(''.join(guess), salt)
                            if crypted == given_hash:
                                print(''.join(guess))
                                exit(0)
                        if len(guess) == 1:
                            guess.append(alphabet[second_letter])
                        else:
                            guess[1] = alphabet[second_letter]
                    if len(guess) == 2:
                        guess.append(alphabet[third_letter])
                    else:
                        guess[2] = alphabet[third_letter]
                if len(guess) == 3:
                    guess.append(alphabet[fourth_letter])
                else:
                    guess[3] = alphabet[fourth_letter]
            if len(guess) == 4:
                guess.append(alphabet[fifth_letter])
            else:
                guess[4] = alphabet[fifth_letter]


if __name__ == "__main__":
    main()

File: pset2/crack/crack.c
//
//      Password cracking program which are hashed with
//      DES algorithm.
//

#include <stdio.h>
#include <cs50.h>
#include <crypt.h>
#include <string.h>
#include <unistd.h>

int main(int argc, string argv[])
{
    // If more than one argument was passed terminate program
    if (argc != 2)
    {
        printf("Usage: ./crack hash\n");
        return 1;
    }

    // Hash passed as argument
    string hash = argv[1];

    // Salt - first two characters of hash
    char salt[3];
    salt[0] = hash[0];
    salt[1] = hash[1];
    salt[2] = '\0';

    // Define and fill guess with \0
    char guess[6];
    for (int i = 1; i <= 5; i++)
    {
        guess[i] = '\0';
    }

    string alphabet = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ";


    // Fift letter loop
    for (int i = 0; i < strlen(alphabet); i++)
    {
        // Fourth letter loop
        for (int j = 0; j < strlen(alphabet); j++)
        {
            // Third letter loop
            for (int k = 0; k < strlen(alphabet); k++)
            {
                // Second letter loop
                for (int l = 0; l < strlen(alphabet); l++)
                {
                    // First letter loop and check if correct
                    for (int m = 0; m < strlen(alphabet); m++)
                    {
                        guess[0] = alphabet[m];
                        string crypted = crypt(guess, salt);
                        // If correct display and quit
                        if (strcmp(hash, crypted) == 0)
                        {
                            printf("%s\n", guess);
                            return 0;
                        }
                    }
                    guess[1] = alphabet[l];
                }
                guess[2] = alphabet[k];
            }
            guess[3] = alphabet[j];
        }
        guess[4] = alphabet[i];
    }
}